In [30]:
#Inizializzazione
import pandas as pd
import numpy as np
import datetime
import commun

# 1.1. Définition de la fonction de nettoyage générique d'un DataFrame
Élimine lignes et colonnes vides et dupliquées


In [31]:
#1.1 Fonction de nettoyage des df

def nettoyage(df_to_clean, label):
    print("Nettoyage du jeu de données: \"%s\" " % label)
    llignes1, lcols1 = df_to_clean.shape
    print(llignes1, lcols1)
    df_to_clean_nac = df_to_clean.dropna(axis=1, how='all')
    llignes2, lcols2 = df_to_clean_nac.shape
    print("Après dropna, cols:", llignes2, lcols2)
    print("Colonnes de moins:\n", lcols1 - lcols2)

    df_to_clean_nacl = df_to_clean_nac.dropna(axis=0, how='all')
    llignes3, lcols3 = df_to_clean_nacl.shape
    print("Après dropna, lignes:", llignes3, lcols3)
    print("Lignes de moins:\n", llignes2 - llignes3)

    df_clean = df_to_clean_nacl.drop_duplicates()
    llignes4, lcols4 = df_clean.shape
    
    pd.set_option('display.float_format', '{:.2f}'.format)
    np.set_printoptions(precision=4)
    dcount_x \
        = pd.DataFrame(df_clean.count()).reset_index().rename(
        columns={'index': 'Indicator', 0: 'count'})
    # print(dcount_x)
    dcount_3 \
        = pd.Series(np.array(
        df_clean.count().tolist())*100/llignes3).rename('percentage')
    dcount_4 \
        = pd.Series(
        llignes3-np.array(df_clean.count().tolist())).rename('vides')
    df_count \
        = dcount_x.merge(dcount_3,left_index=True, right_index=True)
    df_count \
        = df_count.merge(dcount_4, left_index=True, right_index=True)

    print(df_count)

    print("\nRemove duplicates -> Lignes de moins:%i\n"
          % (llignes2 - llignes4))
    # df_clean.shape

    print('Output: ', df_clean.shape)
    print("\n--------------------------------\n")

    return df_clean


# 1.2 Définition de la fonction de fusion de plusieurs DataFrames

In [32]:
def create_df_from_two_dfs_join(l_df, l_merge, how_instr):
    n_joins = len(l_merge)
    print(n_joins, 'mergers to do.\n\n')

    for j in range(n_joins):
        col_merge_on = l_merge[j][2]      
        print('merger n.', j, '(%s)' % col_merge_on)

        if j == 0:
            ind_df1 = l_merge[j][0][0]
            ind_df2 = l_merge[j][1][0]
            
            # df1 = l_df[ind_df1]
            df1 = l_df[ind_df1][l_merge[j][0][1]]
            df1 = df1.astype(str)
            # https://stackoverflow.com/
            # questions/64385747/valueerror-you-are-trying-to-merge-
            # on-object-and-int64-columns-when-use-pandas
            print('DF1:', len(df1.columns.tolist()))
            print(len(l_merge[j][0][1]), df1.columns)
            print('df1.shape before dropna on %s:'
                  % col_merge_on, df1.shape )
            df1 = df1.dropna(axis=0, subset=[col_merge_on])
            print('df1.shape after dropna on %s:'
                  % col_merge_on, df1.shape )
            # df1 = df1.drop_duplicates(subset=col_merge_on)
            df1 = df1.sort_values(by=col_merge_on, ascending=False)
            display(df1.head(2))           
            print('df1[col_merge_on].shape', df1[col_merge_on].shape)

            df2 = l_df[ind_df2][l_merge[j][1][1]]
            df2 = df2.astype(str)
            print('DF2:', len(df2.columns.tolist()))
            print(len(l_merge[j][1][1]), df2.columns)
            print(df2.shape)
            print('df2.shape before dropna on %s:'
                  % col_merge_on, df2.shape)
            df2 = df2.dropna(axis=0, subset=[col_merge_on])
            print('df2.shape after dropna on %s:'
                  % col_merge_on, df2.shape)

            df2 = df2.sort_values(by=col_merge_on, ascending=False)
            display(df2.head(2))
            print('df2[col_merge_on].shape', df2[col_merge_on].shape)
            
            df1 = pd.merge(df1, df2, how=how_instr, on=col_merge_on)

            print('Output of merger n.1', df1.shape)
            print(len(df1.columns.tolist()), df1.columns)
            display(df1.head(2))
            
            print('End of merger 1\n')

        else:
            ind_df2 = l_merge[j][1][0]
            df2 = l_df[ind_df2][l_merge[j][1][1]]
            
            df1 = df1.astype(str)
            # https://stackoverflow.com/questions/64385747/
            # valueerror-you-are-trying-to-merge-on-object
            # -and-int64-columns-when-use-pandas
            print('df1.shape before dropna on %s:'
                  % col_merge_on, df1.shape)
            df1 = df1.dropna(axis=0, subset=[col_merge_on])
            print('df1.shape after dropna on %s:'
                  % col_merge_on, df1.shape)
            df1 = df1.sort_values(by=col_merge_on, ascending=False)          
            print('DF1', df1[col_merge_on].shape, '(df1.head(2))')
            display(df1.head(2))
            
            df2 = df2.astype(str)
            
            print('df2.shape before dropna on %s:'
                  % col_merge_on, df2.shape)
            df2 = df2.dropna(axis=0, subset=[col_merge_on])
            print('df2.shape after dropna on %s:'
                  % col_merge_on, df2.shape)

            df2 = df2.sort_values(by=col_merge_on, ascending=False)
            print('DF2', df2[col_merge_on].shape, '(df2.head(2))')
            display(df2.head(2))
            
            df1 = pd.merge(df1, df2, how=how_instr, on=col_merge_on)

            print('Output of merger n.', j, df1.shape)
            print(len(df1.columns.tolist()), df1.columns)
            display(df1.head(2))
        
    return df1


# 2.1 Importation et caracterisation des jeux de données

In [33]:
l_fichiers = ['olist_customers_dataset', 'olist_geolocation_dataset',
              'olist_order_items_dataset',
              'olist_order_payments_dataset',
              'olist_order_reviews_dataset', 'olist_orders_dataset',
              'olist_products_dataset', 'olist_sellers_dataset',
              'product_category_name_translation']
l_df = []

for i in l_fichiers:
  df = pd.read_csv("data/%s.csv" % i, sep= ',', low_memory=False )
  df = nettoyage(df, str(l_fichiers.index(i)))
  l_df.append(df)


Nettoyage du jeu de données: "0" 
99441 5
Après dropna, cols: 99441 5
Colonnes de moins:
 0
Après dropna, lignes: 99441 5
Lignes de moins:
 0
                  Indicator  count  percentage  vides
0               customer_id  99441      100.00      0
1        customer_unique_id  99441      100.00      0
2  customer_zip_code_prefix  99441      100.00      0
3             customer_city  99441      100.00      0
4            customer_state  99441      100.00      0

Remove duplicates -> Lignes de moins:0

Output:  (99441, 5)

--------------------------------

Nettoyage du jeu de données: "1" 
1000163 5
Après dropna, cols: 1000163 5
Colonnes de moins:
 0
Après dropna, lignes: 1000163 5
Lignes de moins:
 0
                     Indicator   count  percentage   vides
0  geolocation_zip_code_prefix  738332       73.82  261831
1              geolocation_lat  738332       73.82  261831
2              geolocation_lng  738332       73.82  261831
3             geolocation_city  738332       73.82  26

# 2.2 Chiffres clé extraites des données brutes


## 2.2.1 Nombre de clients uniques d'après les données brutes


In [34]:
print('_________________________________\n'
      'Recherche de lignes vides dans le fichier brut')
display(l_df[0][l_df[0].isnull().any(1)])
print((l_df[0].shape))
print((l_df[0].dropna().shape))
print('Aucune ligne vide dans le fichier brut\n '
      '_________________________________\n')

print('_________________________________\n'
      'Recherche de lignes vides dans le fichier brut.'
      'groupby(\'customer_unique_id\').count())')
df1 = l_df[0].groupby('customer_unique_id').count()
print(df1.shape, 'avant dropna')
df1 = df1.dropna()
print(df1.shape, 'après dropna')
print('Aucune ligne vide dans le fichier dans le fichier '
      'brut.groupby(\'customer_unique_id\').count()).\n'
      'C\'est logique puisqu\'il n\'y en avait pas '
      'dans le fichier de depart)'
      '\n _________________________________\n'
      'Le nombre de ligne se reduit à cause de groupby\n'
      'Le nombre de clients uniques est %s' % df1.shape[0])

display(df1.sample(n=2))

_________________________________
Recherche de lignes vides dans le fichier brut


Empty DataFrame
Columns: [customer_id, customer_unique_id, customer_zip_code_prefix, customer_city, customer_state]
Index: []

(99441, 5)
(99441, 5)
Aucune ligne vide dans le fichier brut
 _________________________________

_________________________________
Recherche de lignes vides dans le fichier brut.groupby('customer_unique_id').count())
(96096, 4) avant dropna
(96096, 4) après dropna
Aucune ligne vide dans le fichier dans le fichier brut.groupby('customer_unique_id').count()).
C'est logique puisqu'il n'y en avait pas dans le fichier de depart)
 _________________________________
Le nombre de ligne se reduit à cause de groupby
Le nombre de clients uniques est 96096


customer_id  customer_zip_code_prefix  \
customer_unique_id                                                        
1e8b70cce882badaff158c24c2b9eb24            1                         1   
95ccc510f68e05d7ecceb58a4f37558a            1                         1   

                                  customer_city  customer_state  
customer_unique_id                                               
1e8b70cce882badaff158c24c2b9eb24              1               1  
95ccc510f68e05d7ecceb58a4f37558a              1               1

______________________________________

## 2.2.2 Nombre de clients de retour (données brutes: 3%)

Données brutes

In [35]:
df = l_df[0]
df_gb = pd.DataFrame(df.groupby('customer_unique_id').count())

clienti_tornati_grezzo \
    = pd.DataFrame(df_gb.loc[df_gb['customer_id']>1])
totale_clienti_unici = df_gb.shape[0]
print(totale_clienti_unici, df_gb.columns)

clienti_tornati = clienti_tornati_grezzo
display(clienti_tornati.sample(5))
totale_clienti_tornati = clienti_tornati.shape[0]
print('totale_clienti_tornati', totale_clienti_tornati)
fraction = totale_clienti_tornati*100/totale_clienti_unici
print(fraction)
print('Fraction des clients qui on acheté à nouveau: %.1f'
      % fraction, '%')


96096 Index(['customer_id', 'customer_zip_code_prefix', 'customer_city',
       'customer_state'],
      dtype='object')


customer_id  customer_zip_code_prefix  \
customer_unique_id                                                        
bce006e903be688f122a39f45b210090            2                         2   
82fb46292a999e7db9e2391c407f76e4            2                         2   
6f65f19c2dcfcab8b76b2fafb5b840e1            2                         2   
54732c804725f94304b9f2d1107a3d6e            2                         2   
87a7c3090f96d9da963125a0d7f8193a            2                         2   

                                  customer_city  customer_state  
customer_unique_id                                               
bce006e903be688f122a39f45b210090              2               2  
82fb46292a999e7db9e2391c407f76e4              2               2  
6f65f19c2dcfcab8b76b2fafb5b840e1              2               2  
54732c804725f94304b9f2d1107a3d6e              2               2  
87a7c3090f96d9da963125a0d7f8193a              2               2

totale_clienti_tornati 2997
3.118756243756244
Fraction des clients qui on acheté à nouveau: 3.1 %


# 3. Création du dataframe complet à traiter: df_all_start
> Un df de commandes

In [36]:
l_merge = [[[0, ['customer_id', 'customer_unique_id']],
            [5, ['customer_id', 'order_id', 'order_status',
                 'order_purchase_timestamp',
                 'order_delivered_carrier_date',
                 'order_estimated_delivery_date']],
            'customer_id'],
           [[], [2, ['order_id', 'order_item_id',
                     'product_id', 'price', 'freight_value' ]],
            'order_id'],
           [[], [4, ['order_id','review_score']], 'order_id'],
           [[], [3, ['order_id', 'payment_sequential', 'payment_type',
                     'payment_installments', 'payment_value']],
            'order_id']]

df_merge_1 = create_df_from_two_dfs_join(l_df, l_merge, 'left')
df_all_start = nettoyage(df_merge_1, 'all')

# Pour chaque ligne de la commande order_id,
# on retient seulement la première ligne

df_all_start['dup'] \
    = df_all_start.duplicated(subset='order_id', keep='first')
# très important

4 mergers to do.


merger n. 0 (customer_id)
DF1: 2
2 Index(['customer_id', 'customer_unique_id'], dtype='object')
df1.shape before dropna on customer_id: (99441, 2)
df1.shape after dropna on customer_id: (99441, 2)


customer_id                customer_unique_id
95945  ffffe8b65bbe3087b653a978c870db99  736e6bfa0510aa5b878881a226a5fd89
33870  ffffa3172527f765de70084a7e53aae8  48fd7dec70f2b104a1d5e8c5c639102b

df1[col_merge_on].shape (99441,)
DF2: 6
6 Index(['customer_id', 'order_id', 'order_status', 'order_purchase_timestamp',
       'order_delivered_carrier_date', 'order_estimated_delivery_date'],
      dtype='object')
(99441, 6)
df2.shape before dropna on customer_id: (99441, 6)
df2.shape after dropna on customer_id: (99441, 6)


customer_id                          order_id  \
26445  ffffe8b65bbe3087b653a978c870db99  2e935fa1d39497aa0ec3f1107fbfb5b8   
73902  ffffa3172527f765de70084a7e53aae8  d0e7be325a1c986babc4e1cdb91edc03   

      order_status order_purchase_timestamp order_delivered_carrier_date  \
26445  unavailable      2017-09-29 14:07:03                          nan   
73902    delivered      2017-09-02 11:53:32          2017-09-08 20:04:11   

      order_estimated_delivery_date  
26445           2017-10-24 00:00:00  
73902           2017-09-26 00:00:00

df2[col_merge_on].shape (99441,)
Output of merger n.1 (99441, 7)
7 Index(['customer_id', 'customer_unique_id', 'order_id', 'order_status',
       'order_purchase_timestamp', 'order_delivered_carrier_date',
       'order_estimated_delivery_date'],
      dtype='object')


customer_id                customer_unique_id  \
0  ffffe8b65bbe3087b653a978c870db99  736e6bfa0510aa5b878881a226a5fd89   
1  ffffa3172527f765de70084a7e53aae8  48fd7dec70f2b104a1d5e8c5c639102b   

                           order_id order_status order_purchase_timestamp  \
0  2e935fa1d39497aa0ec3f1107fbfb5b8  unavailable      2017-09-29 14:07:03   
1  d0e7be325a1c986babc4e1cdb91edc03    delivered      2017-09-02 11:53:32   

  order_delivered_carrier_date order_estimated_delivery_date  
0                          nan           2017-10-24 00:00:00  
1          2017-09-08 20:04:11           2017-09-26 00:00:00

End of merger 1

merger n. 1 (order_id)
df1.shape before dropna on order_id: (99441, 7)
df1.shape after dropna on order_id: (99441, 7)
DF1 (99441,) (df1.head(2))


customer_id                customer_unique_id  \
40883  96d649da0cc4ff33bb408b199d4c7dcf  cd76a00d8e3ca5e6ab9ed9ecb6667ac4   
28579  b5e6afd5a41800fdf401e0272ca74655  eb803377c9315b564bdedad672039306   

                               order_id order_status order_purchase_timestamp  \
40883  fffe41c64501cc87c801fd61db3f6244    delivered      2018-06-09 17:00:18   
28579  fffe18544ffabc95dfada21779c9644f    delivered      2017-08-14 23:02:59   

      order_delivered_carrier_date order_estimated_delivery_date  
40883          2018-06-11 14:11:00           2018-06-28 00:00:00  
28579          2017-08-15 19:02:53           2017-08-25 00:00:00

df2.shape before dropna on order_id: (112650, 5)
df2.shape after dropna on order_id: (112650, 5)
DF2 (112650,) (df2.head(2))


order_id order_item_id  \
112649  fffe41c64501cc87c801fd61db3f6244             1   
112648  fffe18544ffabc95dfada21779c9644f             1   

                              product_id  price freight_value  
112649  350688d9dc1e75ff97be326363655e01   43.0         12.79  
112648  9c422a519119dcad7575db5af1ba540e  55.99          8.72

Output of merger n. 1 (113425, 11)
11 Index(['customer_id', 'customer_unique_id', 'order_id', 'order_status',
       'order_purchase_timestamp', 'order_delivered_carrier_date',
       'order_estimated_delivery_date', 'order_item_id', 'product_id', 'price',
       'freight_value'],
      dtype='object')


customer_id                customer_unique_id  \
0  96d649da0cc4ff33bb408b199d4c7dcf  cd76a00d8e3ca5e6ab9ed9ecb6667ac4   
1  b5e6afd5a41800fdf401e0272ca74655  eb803377c9315b564bdedad672039306   

                           order_id order_status order_purchase_timestamp  \
0  fffe41c64501cc87c801fd61db3f6244    delivered      2018-06-09 17:00:18   
1  fffe18544ffabc95dfada21779c9644f    delivered      2017-08-14 23:02:59   

  order_delivered_carrier_date order_estimated_delivery_date order_item_id  \
0          2018-06-11 14:11:00           2018-06-28 00:00:00             1   
1          2017-08-15 19:02:53           2017-08-25 00:00:00             1   

                         product_id  price freight_value  
0  350688d9dc1e75ff97be326363655e01   43.0         12.79  
1  9c422a519119dcad7575db5af1ba540e  55.99          8.72

merger n. 2 (order_id)
df1.shape before dropna on order_id: (113425, 11)
df1.shape after dropna on order_id: (113425, 11)
DF1 (113425,) (df1.head(2))


customer_id                customer_unique_id  \
0  96d649da0cc4ff33bb408b199d4c7dcf  cd76a00d8e3ca5e6ab9ed9ecb6667ac4   
1  b5e6afd5a41800fdf401e0272ca74655  eb803377c9315b564bdedad672039306   

                           order_id order_status order_purchase_timestamp  \
0  fffe41c64501cc87c801fd61db3f6244    delivered      2018-06-09 17:00:18   
1  fffe18544ffabc95dfada21779c9644f    delivered      2017-08-14 23:02:59   

  order_delivered_carrier_date order_estimated_delivery_date order_item_id  \
0          2018-06-11 14:11:00           2018-06-28 00:00:00             1   
1          2017-08-15 19:02:53           2017-08-25 00:00:00             1   

                         product_id  price freight_value  
0  350688d9dc1e75ff97be326363655e01   43.0         12.79  
1  9c422a519119dcad7575db5af1ba540e  55.99          8.72

df2.shape before dropna on order_id: (99224, 2)
df2.shape after dropna on order_id: (99224, 2)
DF2 (99224,) (df2.head(2))


order_id review_score
92457  fffe41c64501cc87c801fd61db3f6244            5
50462  fffe18544ffabc95dfada21779c9644f            5

Output of merger n. 2 (114092, 12)
12 Index(['customer_id', 'customer_unique_id', 'order_id', 'order_status',
       'order_purchase_timestamp', 'order_delivered_carrier_date',
       'order_estimated_delivery_date', 'order_item_id', 'product_id', 'price',
       'freight_value', 'review_score'],
      dtype='object')


customer_id                customer_unique_id  \
0  96d649da0cc4ff33bb408b199d4c7dcf  cd76a00d8e3ca5e6ab9ed9ecb6667ac4   
1  b5e6afd5a41800fdf401e0272ca74655  eb803377c9315b564bdedad672039306   

                           order_id order_status order_purchase_timestamp  \
0  fffe41c64501cc87c801fd61db3f6244    delivered      2018-06-09 17:00:18   
1  fffe18544ffabc95dfada21779c9644f    delivered      2017-08-14 23:02:59   

  order_delivered_carrier_date order_estimated_delivery_date order_item_id  \
0          2018-06-11 14:11:00           2018-06-28 00:00:00             1   
1          2017-08-15 19:02:53           2017-08-25 00:00:00             1   

                         product_id  price freight_value review_score  
0  350688d9dc1e75ff97be326363655e01   43.0         12.79            5  
1  9c422a519119dcad7575db5af1ba540e  55.99          8.72            5

merger n. 3 (order_id)
df1.shape before dropna on order_id: (114092, 12)
df1.shape after dropna on order_id: (114092, 12)
DF1 (114092,) (df1.head(2))


customer_id                customer_unique_id  \
0  96d649da0cc4ff33bb408b199d4c7dcf  cd76a00d8e3ca5e6ab9ed9ecb6667ac4   
1  b5e6afd5a41800fdf401e0272ca74655  eb803377c9315b564bdedad672039306   

                           order_id order_status order_purchase_timestamp  \
0  fffe41c64501cc87c801fd61db3f6244    delivered      2018-06-09 17:00:18   
1  fffe18544ffabc95dfada21779c9644f    delivered      2017-08-14 23:02:59   

  order_delivered_carrier_date order_estimated_delivery_date order_item_id  \
0          2018-06-11 14:11:00           2018-06-28 00:00:00             1   
1          2017-08-15 19:02:53           2017-08-25 00:00:00             1   

                         product_id  price freight_value review_score  
0  350688d9dc1e75ff97be326363655e01   43.0         12.79            5  
1  9c422a519119dcad7575db5af1ba540e  55.99          8.72            5

df2.shape before dropna on order_id: (103886, 5)
df2.shape after dropna on order_id: (103886, 5)
DF2 (103886,) (df2.head(2))


order_id payment_sequential payment_type  \
1235   fffe41c64501cc87c801fd61db3f6244                  1  credit_card   
37089  fffe18544ffabc95dfada21779c9644f                  1  credit_card   

      payment_installments payment_value  
1235                     1         55.79  
37089                    3         64.71

Output of merger n. 3 (119143, 16)
16 Index(['customer_id', 'customer_unique_id', 'order_id', 'order_status',
       'order_purchase_timestamp', 'order_delivered_carrier_date',
       'order_estimated_delivery_date', 'order_item_id', 'product_id', 'price',
       'freight_value', 'review_score', 'payment_sequential', 'payment_type',
       'payment_installments', 'payment_value'],
      dtype='object')


customer_id                customer_unique_id  \
0  96d649da0cc4ff33bb408b199d4c7dcf  cd76a00d8e3ca5e6ab9ed9ecb6667ac4   
1  b5e6afd5a41800fdf401e0272ca74655  eb803377c9315b564bdedad672039306   

                           order_id order_status order_purchase_timestamp  \
0  fffe41c64501cc87c801fd61db3f6244    delivered      2018-06-09 17:00:18   
1  fffe18544ffabc95dfada21779c9644f    delivered      2017-08-14 23:02:59   

  order_delivered_carrier_date order_estimated_delivery_date order_item_id  \
0          2018-06-11 14:11:00           2018-06-28 00:00:00             1   
1          2017-08-15 19:02:53           2017-08-25 00:00:00             1   

                         product_id  price freight_value review_score  \
0  350688d9dc1e75ff97be326363655e01   43.0         12.79            5   
1  9c422a519119dcad7575db5af1ba540e  55.99          8.72            5   

  payment_sequential payment_type payment_installments payment_value  
0                  1  credit_card                    1         55.79  
1                  1  credit_card                    3         64.71

Nettoyage du jeu de données: "all" 
119143 16
Après dropna, cols: 119143 16
Colonnes de moins:
 0
Après dropna, lignes: 119143 16
Lignes de moins:
 0
                        Indicator   count  percentage  vides
0                     customer_id  118722       99.65    421
1              customer_unique_id  118722       99.65    421
2                        order_id  118722       99.65    421
3                    order_status  118722       99.65    421
4        order_purchase_timestamp  118722       99.65    421
5    order_delivered_carrier_date  118722       99.65    421
6   order_estimated_delivery_date  118722       99.65    421
7                   order_item_id  118722       99.65    421
8                      product_id  118722       99.65    421
9                           price  118722       99.65    421
10                  freight_value  118722       99.65    421
11                   review_score  118722       99.65    421
12             payment_sequential  118719       99.64    

C'est-à-dire que 118719 achats de *produits* (les commandes peuvent inclure
plus d'un produit et le meme produit achète plusieurs est toujours
indiqué avec le meme product_id, qui se répètera) sont enregistrés.
Il reste à trier les commandes non livrées.
____________________________________

# 3.1 Nombre de clients uniques de df_all ()

In [37]:
print('Le numéro de clients uniques est confirmé: %s'
      % df_all_start.groupby('customer_unique_id').count().shape[0])

Le numéro de clients uniques est confirmé: 96096


# 3.2 Traitement des dates sur df_all_start -> df_commandes_livrees

In [38]:
def select_livraisons(df):
    status_dict = {
        'approved': 0,
        'canceled': 0,
        'created': 0,
        'delivered': 1,
        'invoiced': 0,
        'processing': 0,
        'shipped': 1,
        'unavailable': 0
}

    df.groupby('order_status').count()
    df.groupby('order_status').count().index

    df['livree'] = df['order_status'].map(status_dict)

    return df

df_commandes_livrees = select_livraisons(df_all_start)

print('Commandes non livrées:',
      df_commandes_livrees.loc[
          df_commandes_livrees['livree']==0].shape[0])
df_commandes_livrees \
    = df_commandes_livrees.loc[df_commandes_livrees['livree']==1]
print('Df des commandes livrées:', df_commandes_livrees.shape)
df_commandes_livrees.columns


Commandes non livrées: 2159
Df des commandes livrées: (116563, 18)


Index(['customer_id', 'customer_unique_id', 'order_id', 'order_status',
       'order_purchase_timestamp', 'order_delivered_carrier_date',
       'order_estimated_delivery_date', 'order_item_id', 'product_id', 'price',
       'freight_value', 'review_score', 'payment_sequential', 'payment_type',
       'payment_installments', 'payment_value', 'dup', 'livree'],
      dtype='object')

***Dates limite***

In [39]:
df = df_commandes_livrees

debut_2017 = "2016-12-31 23:59:59"

#Régularisation du format de la date

for i in ['order_purchase_timestamp', 'order_delivered_carrier_date',
          'order_estimated_delivery_date']:
        df[i] = pd.to_datetime(df[i])
        # https://paritoshsharmaghimire.medium.com/
        # datetime-in-pandas-read-csv-everything-you-have-to-know-8245bd048fa

df = df.sort_values(by='order_purchase_timestamp', ascending=False)

least_recent_date = df['order_purchase_timestamp'].min()
most_recent_date = df['order_purchase_timestamp'].max()

display(df.loc[df['order_delivered_carrier_date']
               ==df['order_delivered_carrier_date'].max()])
print(most_recent_date)
print(most_recent_date.day)

print(least_recent_date)

aujourdhui = datetime.datetime.strptime("2018-09-04 23:59:00",
                                        "%Y-%m-%d %H:%M:%S")

display(df.head(5), df.tail(5))

now_ts = datetime.datetime.strptime("2018-09-03 23:59:00",
                                    "%Y-%m-%d %H:%M:%S")
# fin du jour de la derniere commande livrée

for i in ['order_purchase_timestamp', 'order_delivered_carrier_date',
          'order_estimated_delivery_date']:
        str_recency = 'recency_%s' % i
        df[str_recency] = (now_ts-df[i]).dt.components['days']

df = df.sort_values(by='order_purchase_timestamp', ascending=False)

df['attente'] = df['order_delivered_carrier_date']\
                -df['order_purchase_timestamp']
df['attente-jours'] = df['attente'].dt.components['days']

# Élimination commandes avec une attente negative entre
# achat et reception (erreurs de saisie)
print('Élimination commandes avec une attente negative '
      'entre achat et reception (erreurs de saisie)\n%s ->'
      % str(df.shape))
df = df.loc[df['attente-jours']>=0] #
print(df.shape)

df['decalage-prevision'] = df['order_delivered_carrier_date']\
                           -df['order_estimated_delivery_date']
df['decalage-prevision-jours'] \
    = df['decalage-prevision'].dt.components['days']

# Mesure croissante de la rapidité de la livraison par rapport à la prévision
df['rapidite_livraison'] = df['decalage-prevision-jours'] * (-1) + 99

print(df['recency_order_purchase_timestamp'].max(),
      df['recency_order_purchase_timestamp'].min())

df['recency_crescent'] = df['recency_order_purchase_timestamp']*(-1)\
                         +df['recency_order_purchase_timestamp'].max()

display(df.head(5), df.tail(5))

df_commandes_livrees = df
df_commandes_livrees.columns


/var/folders/zv/tmk76ddj1sj2bmhyt0by55l40000gp/T/ipykernel_1083/1999822649.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[i] = pd.to_datetime(df[i])
/var/folders/zv/tmk76ddj1sj2bmhyt0by55l40000gp/T/ipykernel_1083/1999822649.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[i] = pd.to_datetime(df[i])
/var/folders/zv/tmk76ddj1sj2bmhyt0by55l40000gp/T/ipykernel_1083/1999822649.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexe

customer_id                customer_unique_id  \
86906  a35878bee339b45240b5a327d933509b  9726a24d281565192465aea2248a165c   

                               order_id order_status order_purchase_timestamp  \
86906  4505acb3759da6b9c7d79a80d29ab3bb    delivered      2018-08-06 14:32:27   

      order_delivered_carrier_date order_estimated_delivery_date  \
86906          2018-09-11 19:48:28                    2018-08-17   

      order_item_id                        product_id  price freight_value  \
86906             1  0eb8e03d4ef83152c4642b606b5bbdec  29.99         12.51   

      review_score payment_sequential payment_type payment_installments  \
86906            5                  1       boleto                    1   

      payment_value    dup  livree  
86906          42.5  False       1

2018-09-03 09:06:57
3
2016-09-04 21:15:19


customer_id                customer_unique_id  \
79751   4b7decb9b58e2569548b8b4c8e20e8d7  ff22e30958c13ffe219db7d711e8f564   
94189   898b7fee99c4e42170ab69ba59be0a8b  24ac2b4327e25baf39f2119e4228976a   
117345  496630b6740bcca28fce9ba50d8a26ef  b701bebbdf478f5500348f03aff62121   
108775  6e353700bc7bcdf6ebc15d6de16d7002  7febafa06d9d8f232a900a2937f04338   
113984  e60df9449653a95af4549bbfcb18a6eb  5c58de6fb80e93396e2f35642666b693   

                                order_id order_status  \
79751   54282e97f61c23b78330c15b154c867d      shipped   
94189   35a972d7f8436f405b56e36add1a7140    delivered   
117345  03ef5dedbe7492bdae72eec50764c43f    delivered   
108775  168626408cb32af0ffaf76711caae1dc    delivered   
113984  0b223d92c27432930dfe407c6aea3041    delivered   

       order_purchase_timestamp order_delivered_carrier_date  \
79751       2018-09-03 09:06:57          2018-09-04 15:25:00   
94189       2018-08-29 15:00:37          2018-08-29 16:57:00   
117345      2018-08-29 14:52:00          2018-08-29 20:01:00   
108775      2018-08-29 14:18:28          2018-08-29 18:51:00   
113984      2018-08-29 14:18:23          2018-08-29 15:29:00   

       order_estimated_delivery_date order_item_id  \
79751                     2018-09-06             1   
94189                     2018-09-05             1   
117345                    2018-09-03             1   
108775                    2018-09-11             1   
113984                    2018-09-04             1   

                              product_id  price freight_value review_score  \
79751   b98992ea80b467987a7fbb88e7f2076a  145.0         21.46            1   
94189   d04857e7b4b708ee8b8b9921163edba3  84.99          8.76            5   
117345  c7f27c5bef2338541c772b5776403e6a   24.9          8.33            3   
108775  bdcf6a834e8faa30dac3886c7a58e92e   45.9         15.39            5   
113984  2b4472df15512a2825ae86fd9ae79335  209.0         46.48            5   

       payment_sequential payment_type payment_installments payment_value  \
79751                   1      voucher                    1        166.46   
94189                   1  credit_card                    1         93.75   
117345                  1  credit_card                    1         33.23   
108775                  1      voucher                    1         61.29   
113984                  1  credit_card                    8        510.96   

          dup  livree  
79751   False       1  
94189   False       1  
117345  False       1  
108775  False       1  
113984   True       1

customer_id                customer_unique_id  \
29849  86dc2ffce2dfff336de2f386a786e574  830d5b7aaa3b6f1e9ad63703bec97d23   
29848  86dc2ffce2dfff336de2f386a786e574  830d5b7aaa3b6f1e9ad63703bec97d23   
29847  86dc2ffce2dfff336de2f386a786e574  830d5b7aaa3b6f1e9ad63703bec97d23   
97652  08c5351a6aca1c1589a38f244edeee9d  b7d76e111c89f7ebf14761390f0f7d17   
97653  08c5351a6aca1c1589a38f244edeee9d  b7d76e111c89f7ebf14761390f0f7d17   

                               order_id order_status order_purchase_timestamp  \
29849  bfbd0f9bdef84302105ad712db648a6c    delivered      2016-09-15 12:16:38   
29848  bfbd0f9bdef84302105ad712db648a6c    delivered      2016-09-15 12:16:38   
29847  bfbd0f9bdef84302105ad712db648a6c    delivered      2016-09-15 12:16:38   
97652  2e7a8482f6fb09756ca50c10d7bfc047      shipped      2016-09-04 21:15:19   
97653  2e7a8482f6fb09756ca50c10d7bfc047      shipped      2016-09-04 21:15:19   

      order_delivered_carrier_date order_estimated_delivery_date  \
29849          2016-11-07 17:11:53                    2016-10-04   
29848          2016-11-07 17:11:53                    2016-10-04   
29847          2016-11-07 17:11:53                    2016-10-04   
97652          2016-10-18 13:14:51                    2016-10-20   
97653          2016-10-18 13:14:51                    2016-10-20   

      order_item_id                        product_id  price freight_value  \
29849             1  5a6b04657a4c5ee34285d1e4619a96b4  44.99          2.83   
29848             2  5a6b04657a4c5ee34285d1e4619a96b4  44.99          2.83   
29847             3  5a6b04657a4c5ee34285d1e4619a96b4  44.99          2.83   
97652             1  c1488892604e4ba5cff5b4eb4d595400  39.99         31.67   
97653             2  f293394c72c9b5fafd7023301fc21fc2   32.9         31.67   

      review_score payment_sequential payment_type payment_installments  \
29849            1                NaN          NaN                  NaN   
29848            1                NaN          NaN                  NaN   
29847            1                NaN          NaN                  NaN   
97652            1                  1  credit_card                    1   
97653            1                  1  credit_card                    1   

      payment_value    dup  livree  
29849           NaN   True       1  
29848           NaN   True       1  
29847           NaN  False       1  
97652        136.23  False       1  
97653        136.23   True       1

Élimination commandes avec une attente negative entre achat et reception (erreurs de saisie)
(116563, 23) ->
(116372, 23)
729 0


customer_id                customer_unique_id  \
79751   4b7decb9b58e2569548b8b4c8e20e8d7  ff22e30958c13ffe219db7d711e8f564   
94189   898b7fee99c4e42170ab69ba59be0a8b  24ac2b4327e25baf39f2119e4228976a   
117345  496630b6740bcca28fce9ba50d8a26ef  b701bebbdf478f5500348f03aff62121   
108775  6e353700bc7bcdf6ebc15d6de16d7002  7febafa06d9d8f232a900a2937f04338   
113984  e60df9449653a95af4549bbfcb18a6eb  5c58de6fb80e93396e2f35642666b693   

                                order_id order_status  \
79751   54282e97f61c23b78330c15b154c867d      shipped   
94189   35a972d7f8436f405b56e36add1a7140    delivered   
117345  03ef5dedbe7492bdae72eec50764c43f    delivered   
108775  168626408cb32af0ffaf76711caae1dc    delivered   
113984  0b223d92c27432930dfe407c6aea3041    delivered   

       order_purchase_timestamp order_delivered_carrier_date  \
79751       2018-09-03 09:06:57          2018-09-04 15:25:00   
94189       2018-08-29 15:00:37          2018-08-29 16:57:00   
117345      2018-08-29 14:52:00          2018-08-29 20:01:00   
108775      2018-08-29 14:18:28          2018-08-29 18:51:00   
113984      2018-08-29 14:18:23          2018-08-29 15:29:00   

       order_estimated_delivery_date order_item_id  \
79751                     2018-09-06             1   
94189                     2018-09-05             1   
117345                    2018-09-03             1   
108775                    2018-09-11             1   
113984                    2018-09-04             1   

                              product_id  price  ... livree  \
79751   b98992ea80b467987a7fbb88e7f2076a  145.0  ...      1   
94189   d04857e7b4b708ee8b8b9921163edba3  84.99  ...      1   
117345  c7f27c5bef2338541c772b5776403e6a   24.9  ...      1   
108775  bdcf6a834e8faa30dac3886c7a58e92e   45.9  ...      1   
113984  2b4472df15512a2825ae86fd9ae79335  209.0  ...      1   

       recency_order_purchase_timestamp recency_order_delivered_carrier_date  \
79751                                 0                                -1.00   
94189                                 5                                 5.00   
117345                                5                                 5.00   
108775                                5                                 5.00   
113984                                5                                 5.00   

       recency_order_estimated_delivery_date         attente attente-jours  \
79751                                     -3 1 days 06:18:03          1.00   
94189                                     -2 0 days 01:56:23          0.00   
117345                                     0 0 days 05:09:00          0.00   
108775                                    -8 0 days 04:32:32          0.00   
113984                                    -1 0 days 01:10:37          0.00   

        decalage-prevision  decalage-prevision-jours  rapidite_livraison  \
79751    -2 days +15:25:00                        -2                 101   
94189    -7 days +16:57:00                        -7                 106   
117345   -5 days +20:01:00                        -5                 104   
108775  -13 days +18:51:00                       -13                 112   
113984   -6 days +15:29:00                        -6                 105   

        recency_crescent  
79751                729  
94189                724  
117345               724  
108775               724  
113984               724  

[5 rows x 27 columns]

customer_id                customer_unique_id  \
29849  86dc2ffce2dfff336de2f386a786e574  830d5b7aaa3b6f1e9ad63703bec97d23   
29848  86dc2ffce2dfff336de2f386a786e574  830d5b7aaa3b6f1e9ad63703bec97d23   
29847  86dc2ffce2dfff336de2f386a786e574  830d5b7aaa3b6f1e9ad63703bec97d23   
97652  08c5351a6aca1c1589a38f244edeee9d  b7d76e111c89f7ebf14761390f0f7d17   
97653  08c5351a6aca1c1589a38f244edeee9d  b7d76e111c89f7ebf14761390f0f7d17   

                               order_id order_status order_purchase_timestamp  \
29849  bfbd0f9bdef84302105ad712db648a6c    delivered      2016-09-15 12:16:38   
29848  bfbd0f9bdef84302105ad712db648a6c    delivered      2016-09-15 12:16:38   
29847  bfbd0f9bdef84302105ad712db648a6c    delivered      2016-09-15 12:16:38   
97652  2e7a8482f6fb09756ca50c10d7bfc047      shipped      2016-09-04 21:15:19   
97653  2e7a8482f6fb09756ca50c10d7bfc047      shipped      2016-09-04 21:15:19   

      order_delivered_carrier_date order_estimated_delivery_date  \
29849          2016-11-07 17:11:53                    2016-10-04   
29848          2016-11-07 17:11:53                    2016-10-04   
29847          2016-11-07 17:11:53                    2016-10-04   
97652          2016-10-18 13:14:51                    2016-10-20   
97653          2016-10-18 13:14:51                    2016-10-20   

      order_item_id                        product_id  price  ... livree  \
29849             1  5a6b04657a4c5ee34285d1e4619a96b4  44.99  ...      1   
29848             2  5a6b04657a4c5ee34285d1e4619a96b4  44.99  ...      1   
29847             3  5a6b04657a4c5ee34285d1e4619a96b4  44.99  ...      1   
97652             1  c1488892604e4ba5cff5b4eb4d595400  39.99  ...      1   
97653             2  f293394c72c9b5fafd7023301fc21fc2   32.9  ...      1   

      recency_order_purchase_timestamp recency_order_delivered_carrier_date  \
29849                              718                               665.00   
29848                              718                               665.00   
29847                              718                               665.00   
97652                              729                               685.00   
97653                              729                               685.00   

      recency_order_estimated_delivery_date          attente attente-jours  \
29849                                   699 53 days 04:55:15         53.00   
29848                                   699 53 days 04:55:15         53.00   
29847                                   699 53 days 04:55:15         53.00   
97652                                   683 43 days 15:59:32         43.00   
97653                                   683 43 days 15:59:32         43.00   

       decalage-prevision  decalage-prevision-jours  rapidite_livraison  \
29849    34 days 17:11:53                        34                  65   
29848    34 days 17:11:53                        34                  65   
29847    34 days 17:11:53                        34                  65   
97652   -2 days +13:14:51                        -2                 101   
97653   -2 days +13:14:51                        -2                 101   

       recency_crescent  
29849                11  
29848                11  
29847                11  
97652                 0  
97653                 0  

[5 rows x 27 columns]

Index(['customer_id', 'customer_unique_id', 'order_id', 'order_status',
       'order_purchase_timestamp', 'order_delivered_carrier_date',
       'order_estimated_delivery_date', 'order_item_id', 'product_id', 'price',
       'freight_value', 'review_score', 'payment_sequential', 'payment_type',
       'payment_installments', 'payment_value', 'dup', 'livree',
       'recency_order_purchase_timestamp',
       'recency_order_delivered_carrier_date',
       'recency_order_estimated_delivery_date', 'attente', 'attente-jours',
       'decalage-prevision', 'decalage-prevision-jours', 'rapidite_livraison',
       'recency_crescent'],
      dtype='object')

> Les commandes sont maintenant rangées en ordre chronologique, du plus récent au plus ancien.
On peut eliminer les order_id dupliqués en étant sur.e.s que les commandes retenues seront les plus récentes.

# 3.3 df_commandes_livrees -> df_clients


In [40]:
# En éliminant les order_id dupliqués de df_all,
# on s'approche au df complet des clients uniques.
# Les lignes avec le meme order_id peuvent se répeter dans
# le cas où plusieurs solutions de paiement aient été choisies.
# Toutefois l'élimination des order_id dupliqués conserve l'information
# sur le montant total de la commande qui apparait
# sur chaque ligne. Il est donc correct de supprimer les order_id dupliqués
# afin de ne pas additionner des montants qui se répètent.

df_clients = df_commandes_livrees.drop_duplicates(subset='order_id')

print(len(df_clients.columns))
display('df_clients.shape', df_clients.shape)
df_clients['payment_value'] = df_clients['payment_value'].astype(float)
# Très important

df_clients_sum = df_clients[
    ['customer_unique_id','payment_value']]\
    .groupby('customer_unique_id').sum()
df_clients_sum \
    = df_clients_sum.reset_index().rename(
    columns={'index': 'customer_unique_id'})

display(df_clients_sum.head)
display(df_clients_sum.shape)


27


'df_clients.shape'

(97417, 27)

/var/folders/zv/tmk76ddj1sj2bmhyt0by55l40000gp/T/ipykernel_1083/1243763359.py:14: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_clients['payment_value'] = df_clients['payment_value'].astype(float)


<bound method NDFrame.head of                      customer_unique_id  payment_value
0      0000366f3b9a7992bf8c76cfdf3221e2         141.90
1      0000b849f77a49e4a4ce2b2a4ca5be3f          27.19
2      0000f46a3911fa3c0805444483337064          86.22
3      0000f6ccb0745a6a4b88665a16c9f078          43.62
4      0004aac84e0df4da2b147fca70cf8255         196.89
...                                 ...            ...
94229  fffcf5a5ff07b0908bd4e2dbc735a684        2067.42
94230  fffea47cd6d3cc0a88bd621562a9d061          84.58
94231  ffff371b4d645b6ecea244b27531430a         112.46
94232  ffff5962728ec6157033ef9805bacc48         133.69
94233  ffffd2657e2aad2907e67c3e9daecbeb          71.56

[94234 rows x 2 columns]>

(94234, 2)

In [41]:
l_df_all = [df_clients, df_clients_sum]

l_merge = [[[0, ['customer_id', 'customer_unique_id', 'order_status',
                 'order_id', 'order_purchase_timestamp',
                 'order_delivered_carrier_date',
                 'order_estimated_delivery_date', 'review_score',
                 'livree', 'recency_order_purchase_timestamp',
                 'recency_order_delivered_carrier_date',
                 'recency_order_estimated_delivery_date', 'attente',
                 'attente-jours', 'decalage-prevision',
                 'decalage-prevision-jours', 'rapidite_livraison',
                 'recency_crescent']], # dernier df commandes avec
                                       # les commandes ordonnées
                                       # par date de livraison
                                       # et la date de la dernière
                                       # commande par client
            [1, ['customer_unique_id', 'payment_value']],
            'customer_unique_id']]

df_clients_new = create_df_from_two_dfs_join(l_df_all, l_merge, 'left')

print('df_clients_new.head()')
display(df_clients_new.head())
display('df_clients_new.shape', df_clients_new.shape)

df_clients_new2 = \
    df_clients_new.drop_duplicates(subset='customer_unique_id')

print('df_clients_new2.head()')
display(df_clients_new2.head())
display('df_clients_new2.shape', df_clients_new2.shape)

1 mergers to do.


merger n. 0 (customer_unique_id)
DF1: 18
18 Index(['customer_id', 'customer_unique_id', 'order_status', 'order_id',
       'order_purchase_timestamp', 'order_delivered_carrier_date',
       'order_estimated_delivery_date', 'review_score', 'livree',
       'recency_order_purchase_timestamp',
       'recency_order_delivered_carrier_date',
       'recency_order_estimated_delivery_date', 'attente', 'attente-jours',
       'decalage-prevision', 'decalage-prevision-jours', 'rapidite_livraison',
       'recency_crescent'],
      dtype='object')
df1.shape before dropna on customer_unique_id: (97417, 18)
df1.shape after dropna on customer_unique_id: (97417, 18)


customer_id                customer_unique_id  \
3966   70bc4d9cf2bbeea3f7b09798370f5884  ffffd2657e2aad2907e67c3e9daecbeb   
50097  832a3b0254347d409512ae92eaf154a6  ffff5962728ec6157033ef9805bacc48   

      order_status                          order_id order_purchase_timestamp  \
3966     delivered  f79a35da168301ae56922475da21117b      2017-05-02 20:18:45   
50097    delivered  94d3ee0bc2a0af9d4fa47a4d63616e8d      2018-05-02 15:17:41   

      order_delivered_carrier_date order_estimated_delivery_date review_score  \
3966           2017-05-04 12:05:22           2017-05-26 00:00:00            5   
50097          2018-05-03 16:45:00           2018-05-28 00:00:00            5   

      livree recency_order_purchase_timestamp  \
3966       1                              489   
50097      1                              124   

      recency_order_delivered_carrier_date  \
3966                                 487.0   
50097                                123.0   

      recency_order_estimated_delivery_date          attente attente-jours  \
3966                                    465  1 days 15:46:37           1.0   
50097                                    98  1 days 01:27:19           1.0   

       decalage-prevision decalage-prevision-jours rapidite_livraison  \
3966   -22 days +12:05:22                      -22                121   
50097  -25 days +16:45:00                      -25                124   

      recency_crescent  
3966               240  
50097              605

df1[col_merge_on].shape (97417,)
DF2: 2
2 Index(['customer_unique_id', 'payment_value'], dtype='object')
(94234, 2)
df2.shape before dropna on customer_unique_id: (94234, 2)
df2.shape after dropna on customer_unique_id: (94234, 2)


customer_unique_id payment_value
94233  ffffd2657e2aad2907e67c3e9daecbeb         71.56
94232  ffff5962728ec6157033ef9805bacc48        133.69

df2[col_merge_on].shape (94234,)
Output of merger n.1 (97417, 19)
19 Index(['customer_id', 'customer_unique_id', 'order_status', 'order_id',
       'order_purchase_timestamp', 'order_delivered_carrier_date',
       'order_estimated_delivery_date', 'review_score', 'livree',
       'recency_order_purchase_timestamp',
       'recency_order_delivered_carrier_date',
       'recency_order_estimated_delivery_date', 'attente', 'attente-jours',
       'decalage-prevision', 'decalage-prevision-jours', 'rapidite_livraison',
       'recency_crescent', 'payment_value'],
      dtype='object')


customer_id                customer_unique_id  \
0  70bc4d9cf2bbeea3f7b09798370f5884  ffffd2657e2aad2907e67c3e9daecbeb   
1  832a3b0254347d409512ae92eaf154a6  ffff5962728ec6157033ef9805bacc48   

  order_status                          order_id order_purchase_timestamp  \
0    delivered  f79a35da168301ae56922475da21117b      2017-05-02 20:18:45   
1    delivered  94d3ee0bc2a0af9d4fa47a4d63616e8d      2018-05-02 15:17:41   

  order_delivered_carrier_date order_estimated_delivery_date review_score  \
0          2017-05-04 12:05:22           2017-05-26 00:00:00            5   
1          2018-05-03 16:45:00           2018-05-28 00:00:00            5   

  livree recency_order_purchase_timestamp  \
0      1                              489   
1      1                              124   

  recency_order_delivered_carrier_date recency_order_estimated_delivery_date  \
0                                487.0                                   465   
1                                123.0                                    98   

           attente attente-jours  decalage-prevision decalage-prevision-jours  \
0  1 days 15:46:37           1.0  -22 days +12:05:22                      -22   
1  1 days 01:27:19           1.0  -25 days +16:45:00                      -25   

  rapidite_livraison recency_crescent payment_value  
0                121              240         71.56  
1                124              605        133.69

End of merger 1

df_clients_new.head()


customer_id                customer_unique_id  \
0  70bc4d9cf2bbeea3f7b09798370f5884  ffffd2657e2aad2907e67c3e9daecbeb   
1  832a3b0254347d409512ae92eaf154a6  ffff5962728ec6157033ef9805bacc48   
2  27f584b0f1dc4e610065c240f68b6be0  ffff371b4d645b6ecea244b27531430a   
3  0ecf8e0a08148af0bf313184f167670a  fffea47cd6d3cc0a88bd621562a9d061   
4  74be082247cd677a147d83ee670e9d53  fffcf5a5ff07b0908bd4e2dbc735a684   

  order_status                          order_id order_purchase_timestamp  \
0    delivered  f79a35da168301ae56922475da21117b      2017-05-02 20:18:45   
1    delivered  94d3ee0bc2a0af9d4fa47a4d63616e8d      2018-05-02 15:17:41   
2    delivered  fdc45e6c7555e6cb3cc0daca2557dbe1      2017-02-07 15:49:16   
3    delivered  c71b9252fd7b3b263aaa4cb09319a323      2017-12-10 20:07:56   
4    delivered  725cf8e9c24e679a8a5a32cb92c9ce1e      2017-06-08 21:00:36   

  order_delivered_carrier_date order_estimated_delivery_date review_score  \
0          2017-05-04 12:05:22           2017-05-26 00:00:00            5   
1          2018-05-03 16:45:00           2018-05-28 00:00:00            5   
2          2017-02-08 10:04:37           2017-03-24 00:00:00            5   
3          2017-12-21 16:27:01           2018-01-12 00:00:00            4   
4          2017-06-23 13:03:27           2017-08-02 00:00:00            5   

  livree recency_order_purchase_timestamp  \
0      1                              489   
1      1                              124   
2      1                              573   
3      1                              267   
4      1                              452   

  recency_order_delivered_carrier_date recency_order_estimated_delivery_date  \
0                                487.0                                   465   
1                                123.0                                    98   
2                                572.0                                   528   
3                                256.0                                   234   
4                                437.0                                   397   

            attente attente-jours  decalage-prevision  \
0   1 days 15:46:37           1.0  -22 days +12:05:22   
1   1 days 01:27:19           1.0  -25 days +16:45:00   
2   0 days 18:15:21           0.0  -44 days +10:04:37   
3  10 days 20:19:05          10.0  -22 days +16:27:01   
4  14 days 16:02:51          14.0  -40 days +13:03:27   

  decalage-prevision-jours rapidite_livraison recency_crescent payment_value  
0                      -22                121              240         71.56  
1                      -25                124              605        133.69  
2                      -44                143              156        112.46  
3                      -22                121              462         84.58  
4                      -40                139              277       2067.42

'df_clients_new.shape'

(97417, 19)

df_clients_new2.head()


customer_id                customer_unique_id  \
0  70bc4d9cf2bbeea3f7b09798370f5884  ffffd2657e2aad2907e67c3e9daecbeb   
1  832a3b0254347d409512ae92eaf154a6  ffff5962728ec6157033ef9805bacc48   
2  27f584b0f1dc4e610065c240f68b6be0  ffff371b4d645b6ecea244b27531430a   
3  0ecf8e0a08148af0bf313184f167670a  fffea47cd6d3cc0a88bd621562a9d061   
4  74be082247cd677a147d83ee670e9d53  fffcf5a5ff07b0908bd4e2dbc735a684   

  order_status                          order_id order_purchase_timestamp  \
0    delivered  f79a35da168301ae56922475da21117b      2017-05-02 20:18:45   
1    delivered  94d3ee0bc2a0af9d4fa47a4d63616e8d      2018-05-02 15:17:41   
2    delivered  fdc45e6c7555e6cb3cc0daca2557dbe1      2017-02-07 15:49:16   
3    delivered  c71b9252fd7b3b263aaa4cb09319a323      2017-12-10 20:07:56   
4    delivered  725cf8e9c24e679a8a5a32cb92c9ce1e      2017-06-08 21:00:36   

  order_delivered_carrier_date order_estimated_delivery_date review_score  \
0          2017-05-04 12:05:22           2017-05-26 00:00:00            5   
1          2018-05-03 16:45:00           2018-05-28 00:00:00            5   
2          2017-02-08 10:04:37           2017-03-24 00:00:00            5   
3          2017-12-21 16:27:01           2018-01-12 00:00:00            4   
4          2017-06-23 13:03:27           2017-08-02 00:00:00            5   

  livree recency_order_purchase_timestamp  \
0      1                              489   
1      1                              124   
2      1                              573   
3      1                              267   
4      1                              452   

  recency_order_delivered_carrier_date recency_order_estimated_delivery_date  \
0                                487.0                                   465   
1                                123.0                                    98   
2                                572.0                                   528   
3                                256.0                                   234   
4                                437.0                                   397   

            attente attente-jours  decalage-prevision  \
0   1 days 15:46:37           1.0  -22 days +12:05:22   
1   1 days 01:27:19           1.0  -25 days +16:45:00   
2   0 days 18:15:21           0.0  -44 days +10:04:37   
3  10 days 20:19:05          10.0  -22 days +16:27:01   
4  14 days 16:02:51          14.0  -40 days +13:03:27   

  decalage-prevision-jours rapidite_livraison recency_crescent payment_value  
0                      -22                121              240         71.56  
1                      -25                124              605        133.69  
2                      -44                143              156        112.46  
3                      -22                121              462         84.58  
4                      -40                139              277       2067.42

'df_clients_new2.shape'

(94234, 19)

# 4. Chiffres clé sur le df des clients des commandes livrées

## 4.1 Chiffre d'affaire des commandes livrées
> Les doublons sur les paiements des ordres sont maintenant éliminés et on peut calculer le chiffre d\'affaire relatif aux commandes livrées.

In [42]:
chiffre_d_affaires = df_clients_new2['payment_value'].astype(float).sum()
print('CA commandes livrées:', chiffre_d_affaires)

CA commandes livrées: 15316365.56


## 4.2 Clients de retour sur les commandes livrées
### 4.2.1 Nombre et fraction sur le total

In [43]:
df_clients = df_clients_new
df_clients_count = pd.DataFrame\
    (df_clients.groupby('customer_unique_id').count())
df_clients_count = df_clients_count.reset_index().\
    rename(columns={'index': 'customer_unique_id'})
print(len(df_clients_count.columns))
display(df_clients_count.head(2))

19


customer_unique_id  customer_id  order_status  order_id  \
0  0000366f3b9a7992bf8c76cfdf3221e2            1             1         1   
1  0000b849f77a49e4a4ce2b2a4ca5be3f            1             1         1   

   order_purchase_timestamp  order_delivered_carrier_date  \
0                         1                             1   
1                         1                             1   

   order_estimated_delivery_date  review_score  livree  \
0                              1             1       1   
1                              1             1       1   

   recency_order_purchase_timestamp  recency_order_delivered_carrier_date  \
0                                 1                                     1   
1                                 1                                     1   

   recency_order_estimated_delivery_date  attente  attente-jours  \
0                                      1        1              1   
1                                      1        1              1   

   decalage-prevision  decalage-prevision-jours  rapidite_livraison  \
0                   1                         1                   1   
1                   1                         1                   1   

   recency_crescent  payment_value  
0                 1              1  
1                 1              1

In [44]:
# C.A. clients qui retournent

totale_clienti_unici = df_clients_count.shape[0]
print('totale_clienti_unici', totale_clienti_unici,
      df_clients_count.columns)

print('clienti tornati')
clienti_tornati_count = pd.DataFrame\
    (df_clients_count.loc[df_clients_count['customer_id']>1])

display(clienti_tornati_count.head(2))
display(clienti_tornati_count.tail(2))

totale_clienti_tornati_count = clienti_tornati_count.shape[0]
print('totale_clienti_tornati_count', totale_clienti_tornati_count)
fraction = totale_clienti_tornati_count*100/totale_clienti_unici
print(fraction)
print('Fraction des clients qui on acheté à nouveau: %.1f'
      % fraction, '%')

totale_clienti_unici 94234 Index(['customer_unique_id', 'customer_id', 'order_status', 'order_id',
       'order_purchase_timestamp', 'order_delivered_carrier_date',
       'order_estimated_delivery_date', 'review_score', 'livree',
       'recency_order_purchase_timestamp',
       'recency_order_delivered_carrier_date',
       'recency_order_estimated_delivery_date', 'attente', 'attente-jours',
       'decalage-prevision', 'decalage-prevision-jours', 'rapidite_livraison',
       'recency_crescent', 'payment_value'],
      dtype='object')
clienti tornati


customer_unique_id  customer_id  order_status  order_id  \
106  004288347e5e88a27ded2bb23747066c            2             2         2   
123  004b45ec5c64187465168251cd1c9c2f            2             2         2   

     order_purchase_timestamp  order_delivered_carrier_date  \
106                         2                             2   
123                         2                             2   

     order_estimated_delivery_date  review_score  livree  \
106                              2             2       2   
123                              2             2       2   

     recency_order_purchase_timestamp  recency_order_delivered_carrier_date  \
106                                 2                                     2   
123                                 2                                     2   

     recency_order_estimated_delivery_date  attente  attente-jours  \
106                                      2        2              2   
123                                      2        2              2   

     decalage-prevision  decalage-prevision-jours  rapidite_livraison  \
106                   2                         2                   2   
123                   2                         2                   2   

     recency_crescent  payment_value  
106                 2              2  
123                 2              2

customer_unique_id  customer_id  order_status  order_id  \
94077  ff922bdd6bafcdf99cb90d7f39cea5b3            3             3         3   
94189  ffe254cc039740e17dd15a5305035928            2             2         2   

       order_purchase_timestamp  order_delivered_carrier_date  \
94077                         3                             3   
94189                         2                             2   

       order_estimated_delivery_date  review_score  livree  \
94077                              3             3       3   
94189                              2             2       2   

       recency_order_purchase_timestamp  recency_order_delivered_carrier_date  \
94077                                 3                                     3   
94189                                 2                                     2   

       recency_order_estimated_delivery_date  attente  attente-jours  \
94077                                      3        3              3   
94189                                      2        2              2   

       decalage-prevision  decalage-prevision-jours  rapidite_livraison  \
94077                   3                         3                   3   
94189                   2                         2                   2   

       recency_crescent  payment_value  
94077                 3              3  
94189                 2              2

totale_clienti_tornati_count 2859
3.0339367956363943
Fraction des clients qui on acheté à nouveau: 3.0 %


### 4.2.1 Chiffre d'affaire rattaché aux clients de retour

In [45]:
# À lancer juste après la case précédente

clienti_tornati_count_l = clienti_tornati_count.index.to_list()

display(clienti_tornati_count.sample(3))

l_df = [clienti_tornati_count, df_clients_new2]
l_merge = [[[0, ['customer_unique_id']],
            [1, ['customer_unique_id', 'order_id', 'payment_value']],
            'customer_unique_id']]

df_ca_ritorno = create_df_from_two_dfs_join(l_df, l_merge, 'left')

df_ca_ritorno = nettoyage(df_ca_ritorno, 'df_ca_ritorno')

display(df_ca_ritorno.shape)
df_ca_ritorno[['customer_unique_id','order_id_x']] \
    = df_ca_ritorno[['customer_unique_id','order_id']].astype(str)
df_ca_ritorno = df_ca_ritorno.drop_duplicates(subset='order_id')
display(df_ca_ritorno.shape)

df_ca_ritorno['payment_value'] \
    = df_ca_ritorno['payment_value'].astype(float)
df_ca_ritorno2 = df_ca_ritorno.groupby('customer_unique_id').sum()
df_ca_ritorno2 = df_ca_ritorno2.reset_index()\
    .rename(columns={'index': 'customer_unique_id'})

df_ca_ritorno2_l = df_ca_ritorno2['customer_unique_id'].to_list()

print(len(clienti_tornati_count_l))
print(len(df_ca_ritorno2_l))
display(df_ca_ritorno2.head(2))
display(df_ca_ritorno2.tail(2))
display(df_ca_ritorno2.shape)

ca_clienti_tornati_count = df_ca_ritorno2['payment_value'].sum()
print('ca_clienti_tornati', ca_clienti_tornati_count)

fraction = ca_clienti_tornati_count*100/chiffre_d_affaires
print(fraction)
print('Fraction du chiffre d\'affaires relatif '
      'aux clients qui ont acheté à nouveau: %.1f' % fraction, '%')

customer_unique_id  customer_id  order_status  order_id  \
16621  2d0aa39055b967248bf551e220b55f59            2             2         2   
78666  d5419994a02a55d4b7cd00e77fabdbae            2             2         2   
1103   02fbe4db886eb98774c8ba300d39fa79            2             2         2   

       order_purchase_timestamp  order_delivered_carrier_date  \
16621                         2                             2   
78666                         2                             2   
1103                          2                             2   

       order_estimated_delivery_date  review_score  livree  \
16621                              2             2       2   
78666                              2             2       2   
1103                               2             2       2   

       recency_order_purchase_timestamp  recency_order_delivered_carrier_date  \
16621                                 2                                     2   
78666                                 2                                     2   
1103                                  2                                     2   

       recency_order_estimated_delivery_date  attente  attente-jours  \
16621                                      2        2              2   
78666                                      2        2              2   
1103                                       2        2              2   

       decalage-prevision  decalage-prevision-jours  rapidite_livraison  \
16621                   2                         2                   2   
78666                   2                         2                   2   
1103                    2                         2                   2   

       recency_crescent  payment_value  
16621                 2              2  
78666                 2              2  
1103                  2              2

1 mergers to do.


merger n. 0 (customer_unique_id)
DF1: 1
1 Index(['customer_unique_id'], dtype='object')
df1.shape before dropna on customer_unique_id: (2859, 1)
df1.shape after dropna on customer_unique_id: (2859, 1)


customer_unique_id
94189  ffe254cc039740e17dd15a5305035928
94077  ff922bdd6bafcdf99cb90d7f39cea5b3

df1[col_merge_on].shape (2859,)
DF2: 3
3 Index(['customer_unique_id', 'order_id', 'payment_value'], dtype='object')
(94234, 3)
df2.shape before dropna on customer_unique_id: (94234, 3)
df2.shape after dropna on customer_unique_id: (94234, 3)


customer_unique_id                          order_id  \
0  ffffd2657e2aad2907e67c3e9daecbeb  f79a35da168301ae56922475da21117b   
1  ffff5962728ec6157033ef9805bacc48  94d3ee0bc2a0af9d4fa47a4d63616e8d   

  payment_value  
0         71.56  
1        133.69

df2[col_merge_on].shape (94234,)
Output of merger n.1 (2859, 3)
3 Index(['customer_unique_id', 'order_id', 'payment_value'], dtype='object')


customer_unique_id                          order_id  \
0  ffe254cc039740e17dd15a5305035928  ebe76afae9242fc70b0b807b78f17adb   
1  ff922bdd6bafcdf99cb90d7f39cea5b3  5f79e2da24a52e76bc22897df03a7901   

  payment_value  
0         80.12  
1         139.6

End of merger 1

Nettoyage du jeu de données: "df_ca_ritorno" 
2859 3
Après dropna, cols: 2859 3
Colonnes de moins:
 0
Après dropna, lignes: 2859 3
Lignes de moins:
 0
            Indicator  count  percentage  vides
0  customer_unique_id   2859      100.00      0
1            order_id   2859      100.00      0
2       payment_value   2859      100.00      0

Remove duplicates -> Lignes de moins:0

Output:  (2859, 3)

--------------------------------



(2859, 3)

(2859, 4)

2859
2859


customer_unique_id  payment_value
0  004288347e5e88a27ded2bb23747066c         354.37
1  004b45ec5c64187465168251cd1c9c2f         147.72

customer_unique_id  payment_value
2857  ff922bdd6bafcdf99cb90d7f39cea5b3         139.60
2858  ffe254cc039740e17dd15a5305035928          80.12

(2859, 2)

ca_clienti_tornati 864141.99
5.641951980153704
Fraction du chiffre d'affaires relatif aux clients qui ont acheté à nouveau: 5.6 %


In [46]:
# Vérification que le df contient des commandes de un ou plusieurs objects
display(df_clients_count.groupby('order_id').count().head())


customer_unique_id  customer_id  order_status  \
order_id                                                  
1                      91375        91375         91375   
2                       2627         2627          2627   
3                        185          185           185   
4                         28           28            28   
5                          9            9             9   

          order_purchase_timestamp  order_delivered_carrier_date  \
order_id                                                           
1                            91375                         91375   
2                             2627                          2627   
3                              185                           185   
4                               28                            28   
5                                9                             9   

          order_estimated_delivery_date  review_score  livree  \
order_id                                                        
1                                 91375         91375   91375   
2                                  2627          2627    2627   
3                                   185           185     185   
4                                    28            28      28   
5                                     9             9       9   

          recency_order_purchase_timestamp  \
order_id                                     
1                                    91375   
2                                     2627   
3                                      185   
4                                       28   
5                                        9   

          recency_order_delivered_carrier_date  \
order_id                                         
1                                        91375   
2                                         2627   
3                                          185   
4                                           28   
5                                            9   

          recency_order_estimated_delivery_date  attente  attente-jours  \
order_id                                                                  
1                                         91375    91375          91375   
2                                          2627     2627           2627   
3                                           185      185            185   
4                                            28       28             28   
5                                             9        9              9   

          decalage-prevision  decalage-prevision-jours  rapidite_livraison  \
order_id                                                                     
1                      91375                     91375               91375   
2                       2627                      2627                2627   
3                        185                       185                 185   
4                         28                        28                  28   
5                          9                         9                   9   

          recency_crescent  payment_value  
order_id                                   
1                    91375          91375  
2                     2627           2627  
3                      185            185  
4                       28             28  
5                        9              9

# 5. Outliers
Si on fixe un seuil pour les paiements par client, les clients outliers sont:

In [47]:
df = df_clients

for i in (10000, 8000, 6000, 4000, 3000):
    df_outliers = pd.DataFrame\
        (df.loc[df['payment_value'].astype(float)>float(i)])
    ca_seuil = df_outliers['payment_value'].astype(float).sum()
    print('seuil:', i, 'réais ~', i*0.195, 'euros \t ->  outliers:',
          df_outliers.shape[0], 'soit', int(ca_seuil), 'BRL (',
          np.around(ca_seuil*100/chiffre_d_affaires, decimals=1),
          '% du CA )')

seuil: 10000 réais ~ 1950.0 euros 	 ->  outliers: 1 soit 13664 BRL ( 0.1 % du CA )
seuil: 8000 réais ~ 1560.0 euros 	 ->  outliers: 1 soit 13664 BRL ( 0.1 % du CA )
seuil: 6000 réais ~ 1170.0 euros 	 ->  outliers: 8 soit 62741 BRL ( 0.4 % du CA )
seuil: 4000 réais ~ 780.0 euros 	 ->  outliers: 20 soit 116168 BRL ( 0.8 % du CA )
seuil: 3000 réais ~ 585.0 euros 	 ->  outliers: 52 soit 221666 BRL ( 1.4 % du CA )


# 6. Exportation

In [48]:
l_df_all = [df_clients_new2, df_clients_count]
# for i in l_df_all:
#     # print(str(i))
#     display(i.columns, i.head(1))

l_merge = [[[0, ['customer_id', 'customer_unique_id', 'order_status',
       'order_purchase_timestamp', 'order_delivered_carrier_date',
       'order_estimated_delivery_date', 'review_score',
       'livree', 'recency_order_purchase_timestamp',
       'recency_order_delivered_carrier_date',
       'recency_order_estimated_delivery_date', 'attente',
       'attente-jours', 'decalage-prevision',
       'decalage-prevision-jours', 'rapidite_livraison',
       'recency_crescent', 'payment_value']],
       [1, ['customer_unique_id', 'order_id'], 'customer_unique_id'],
            'customer_unique_id']]

df_all2 = create_df_from_two_dfs_join(l_df_all, l_merge, 'left')
df_all2 = df_all2.rename(columns={'order_id': 'orders'})

display('head', df_all2.head(), 'tail', df_all2.tail())

display(df_all2.columns)
display(df_all2.shape)


1 mergers to do.


merger n. 0 (customer_unique_id)
DF1: 18
18 Index(['customer_id', 'customer_unique_id', 'order_status',
       'order_purchase_timestamp', 'order_delivered_carrier_date',
       'order_estimated_delivery_date', 'review_score', 'livree',
       'recency_order_purchase_timestamp',
       'recency_order_delivered_carrier_date',
       'recency_order_estimated_delivery_date', 'attente', 'attente-jours',
       'decalage-prevision', 'decalage-prevision-jours', 'rapidite_livraison',
       'recency_crescent', 'payment_value'],
      dtype='object')
df1.shape before dropna on customer_unique_id: (94234, 18)
df1.shape after dropna on customer_unique_id: (94234, 18)


customer_id                customer_unique_id  \
0  70bc4d9cf2bbeea3f7b09798370f5884  ffffd2657e2aad2907e67c3e9daecbeb   
1  832a3b0254347d409512ae92eaf154a6  ffff5962728ec6157033ef9805bacc48   

  order_status order_purchase_timestamp order_delivered_carrier_date  \
0    delivered      2017-05-02 20:18:45          2017-05-04 12:05:22   
1    delivered      2018-05-02 15:17:41          2018-05-03 16:45:00   

  order_estimated_delivery_date review_score livree  \
0           2017-05-26 00:00:00            5      1   
1           2018-05-28 00:00:00            5      1   

  recency_order_purchase_timestamp recency_order_delivered_carrier_date  \
0                              489                                487.0   
1                              124                                123.0   

  recency_order_estimated_delivery_date          attente attente-jours  \
0                                   465  1 days 15:46:37           1.0   
1                                    98  1 days 01:27:19           1.0   

   decalage-prevision decalage-prevision-jours rapidite_livraison  \
0  -22 days +12:05:22                      -22                121   
1  -25 days +16:45:00                      -25                124   

  recency_crescent payment_value  
0              240         71.56  
1              605        133.69

df1[col_merge_on].shape (94234,)
DF2: 2
2 Index(['customer_unique_id', 'order_id'], dtype='object')
(94234, 2)
df2.shape before dropna on customer_unique_id: (94234, 2)
df2.shape after dropna on customer_unique_id: (94234, 2)


customer_unique_id order_id
94233  ffffd2657e2aad2907e67c3e9daecbeb        1
94232  ffff5962728ec6157033ef9805bacc48        1

df2[col_merge_on].shape (94234,)
Output of merger n.1 (94234, 19)
19 Index(['customer_id', 'customer_unique_id', 'order_status',
       'order_purchase_timestamp', 'order_delivered_carrier_date',
       'order_estimated_delivery_date', 'review_score', 'livree',
       'recency_order_purchase_timestamp',
       'recency_order_delivered_carrier_date',
       'recency_order_estimated_delivery_date', 'attente', 'attente-jours',
       'decalage-prevision', 'decalage-prevision-jours', 'rapidite_livraison',
       'recency_crescent', 'payment_value', 'order_id'],
      dtype='object')


customer_id                customer_unique_id  \
0  70bc4d9cf2bbeea3f7b09798370f5884  ffffd2657e2aad2907e67c3e9daecbeb   
1  832a3b0254347d409512ae92eaf154a6  ffff5962728ec6157033ef9805bacc48   

  order_status order_purchase_timestamp order_delivered_carrier_date  \
0    delivered      2017-05-02 20:18:45          2017-05-04 12:05:22   
1    delivered      2018-05-02 15:17:41          2018-05-03 16:45:00   

  order_estimated_delivery_date review_score livree  \
0           2017-05-26 00:00:00            5      1   
1           2018-05-28 00:00:00            5      1   

  recency_order_purchase_timestamp recency_order_delivered_carrier_date  \
0                              489                                487.0   
1                              124                                123.0   

  recency_order_estimated_delivery_date          attente attente-jours  \
0                                   465  1 days 15:46:37           1.0   
1                                    98  1 days 01:27:19           1.0   

   decalage-prevision decalage-prevision-jours rapidite_livraison  \
0  -22 days +12:05:22                      -22                121   
1  -25 days +16:45:00                      -25                124   

  recency_crescent payment_value order_id  
0              240         71.56        1  
1              605        133.69        1

End of merger 1



'head'

customer_id                customer_unique_id  \
0  70bc4d9cf2bbeea3f7b09798370f5884  ffffd2657e2aad2907e67c3e9daecbeb   
1  832a3b0254347d409512ae92eaf154a6  ffff5962728ec6157033ef9805bacc48   
2  27f584b0f1dc4e610065c240f68b6be0  ffff371b4d645b6ecea244b27531430a   
3  0ecf8e0a08148af0bf313184f167670a  fffea47cd6d3cc0a88bd621562a9d061   
4  74be082247cd677a147d83ee670e9d53  fffcf5a5ff07b0908bd4e2dbc735a684   

  order_status order_purchase_timestamp order_delivered_carrier_date  \
0    delivered      2017-05-02 20:18:45          2017-05-04 12:05:22   
1    delivered      2018-05-02 15:17:41          2018-05-03 16:45:00   
2    delivered      2017-02-07 15:49:16          2017-02-08 10:04:37   
3    delivered      2017-12-10 20:07:56          2017-12-21 16:27:01   
4    delivered      2017-06-08 21:00:36          2017-06-23 13:03:27   

  order_estimated_delivery_date review_score livree  \
0           2017-05-26 00:00:00            5      1   
1           2018-05-28 00:00:00            5      1   
2           2017-03-24 00:00:00            5      1   
3           2018-01-12 00:00:00            4      1   
4           2017-08-02 00:00:00            5      1   

  recency_order_purchase_timestamp recency_order_delivered_carrier_date  \
0                              489                                487.0   
1                              124                                123.0   
2                              573                                572.0   
3                              267                                256.0   
4                              452                                437.0   

  recency_order_estimated_delivery_date           attente attente-jours  \
0                                   465   1 days 15:46:37           1.0   
1                                    98   1 days 01:27:19           1.0   
2                                   528   0 days 18:15:21           0.0   
3                                   234  10 days 20:19:05          10.0   
4                                   397  14 days 16:02:51          14.0   

   decalage-prevision decalage-prevision-jours rapidite_livraison  \
0  -22 days +12:05:22                      -22                121   
1  -25 days +16:45:00                      -25                124   
2  -44 days +10:04:37                      -44                143   
3  -22 days +16:27:01                      -22                121   
4  -40 days +13:03:27                      -40                139   

  recency_crescent payment_value orders  
0              240         71.56      1  
1              605        133.69      1  
2              156        112.46      1  
3              462         84.58      1  
4              277       2067.42      1

'tail'

customer_id                customer_unique_id  \
94229  47227568b10f5f58a524a75507e6992c  0004aac84e0df4da2b147fca70cf8255   
94230  914991f0c02ef0843c0e7010c819d642  0000f6ccb0745a6a4b88665a16c9f078   
94231  9b3932a6253894a02c1df9d19004239f  0000f46a3911fa3c0805444483337064   
94232  4cb282e167ae9234755102258dd52ee8  0000b849f77a49e4a4ce2b2a4ca5be3f   
94233  fadbb3709178fc513abc1b2670aa1ad2  0000366f3b9a7992bf8c76cfdf3221e2   

      order_status order_purchase_timestamp order_delivered_carrier_date  \
94229    delivered      2017-11-14 19:45:42          2017-11-16 19:52:10   
94230    delivered      2017-10-12 20:29:41          2017-10-13 20:08:19   
94231    delivered      2017-03-10 21:05:03          2017-03-13 12:58:30   
94232    delivered      2018-05-07 11:11:27          2018-05-09 12:18:00   
94233    delivered      2018-05-10 10:56:27          2018-05-12 08:18:00   

      order_estimated_delivery_date review_score livree  \
94229           2017-12-05 00:00:00            5      1   
94230           2017-11-13 00:00:00            4      1   
94231           2017-04-07 00:00:00            3      1   
94232           2018-05-15 00:00:00            4      1   
94233           2018-05-21 00:00:00            5      1   

      recency_order_purchase_timestamp recency_order_delivered_carrier_date  \
94229                              293                                291.0   
94230                              326                                325.0   
94231                              542                                539.0   
94232                              119                                117.0   
94233                              116                                114.0   

      recency_order_estimated_delivery_date          attente attente-jours  \
94229                                   272  2 days 00:06:28           2.0   
94230                                   294  0 days 23:38:38           0.0   
94231                                   514  2 days 15:53:27           2.0   
94232                                   111  2 days 01:06:33           2.0   
94233                                   105  1 days 21:21:33           1.0   

       decalage-prevision decalage-prevision-jours rapidite_livraison  \
94229  -19 days +19:52:10                      -19                118   
94230  -31 days +20:08:19                      -31                130   
94231  -25 days +12:58:30                      -25                124   
94232   -6 days +12:18:00                       -6                105   
94233   -9 days +08:18:00                       -9                108   

      recency_crescent payment_value orders  
94229              436        196.89      1  
94230              403         43.62      1  
94231              187         86.22      1  
94232              610         27.19      1  
94233              613         141.9      1

Index(['customer_id', 'customer_unique_id', 'order_status',
       'order_purchase_timestamp', 'order_delivered_carrier_date',
       'order_estimated_delivery_date', 'review_score', 'livree',
       'recency_order_purchase_timestamp',
       'recency_order_delivered_carrier_date',
       'recency_order_estimated_delivery_date', 'attente', 'attente-jours',
       'decalage-prevision', 'decalage-prevision-jours', 'rapidite_livraison',
       'recency_crescent', 'payment_value', 'orders'],
      dtype='object')

(94234, 19)

***Controle des valeurs maximales***

In [49]:
df = df_all2
df['payment_value'] = df['payment_value'].astype(float)
print(df['payment_value'].astype(float).max())

df_s_outliers_3000 = df.loc[df['payment_value'].astype(float)<3000]
df_s_outliers_3000_only\
    = df.loc[df['payment_value'].astype(float)>=3000]
print(df_s_outliers_3000['payment_value'].max())


13664.08
2960.05


***EXPORT***

In [50]:
l_write_csv = [[df_s_outliers_3000_only,
                "data/df_olist_clean_outliers_3000_only.csv"]]
# , [df_s_outliers_6000, "data/df_olist_clean_seuil_tot_6000_new.csv"]]

for go_for in range(len(l_write_csv)):
    df = l_write_csv[go_for][0]
    csv_name = l_write_csv[go_for][1]

    # print(df, csv_name)
    df.to_csv(csv_name, sep=',')
